### PR TITLE
SNS support for wallet

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/25907

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Run with `--enable-features=BraveWalletSns`
By now works correctly with `https://solana-api.projectserum.com/` set as Solana rpc endpoint. Default  `https://mainnet-beta-solana.brave.com/rpc` endpoint will get SNS support soon.
For Solana network in wallet's `send to` input
`onybose.sol` resolves to `5wytVPbjLb2VCXbynhUQabEZZD2B6Wxrkvwm6v6Cuy5X` (SOL record on domain)
`brave.onybose.sol` resolves to `3nCNLXJw9FEXgByDLffv2mC9233EdFHnoJx2MySRrhuR` (SOL record on subdomain)
`bonfida.sol` resolves to `HKKp49qGWXd639QsuH7JiLijfVW5UtCVY4s1n2HANwEA` (NFT domain)

### CODE REVIEW

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+
+  auto address =
+      SolanaKeyring::FindProgramDerivedAddress(seeds, kNameProgramId);
+  if (!address)
+    return absl::nullopt;
+
+  return SolanaAddress::FromBase58(*address);
+}
+
+absl::optional<SolanaAddress> GetMintAddress(
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:264]: Link to https://github.com/Bonfida/name-tokenizer#mint

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+
+  auto address =
+      SolanaKeyring::FindProgramDerivedAddress(seeds, kNameTokenizerId);
+  if (!address)
+    return absl::nullopt;
+
+  return SolanaAddress::FromBase58(*address);
+}
+
+absl::optional<SolanaAddress> GetDomainKey(const std::string& domain,
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:285]: Link to https://github.com/Bonfida/solana-program-library/blob/6e3be3eedad3a7f4a83c1b7cd5f17f89231e0bca/name-service/js/src/utils.ts#L158

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+      }
+    }
+
+    // Check if domain is minted and its supply is 1.
+    if (!nft_mint_supply_check_done_) {
+      FetchNftSplMint();
+      return;
+    }
+
+    // Find token account for that mint and extract owner from its data.
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:459]: // Find token account for that mint and extract owner from its data to be the resolution result.

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+  ScopedWorkOnSnsTask work_on_task(this);
+
+  if (!api_request_result.Is2XXResponseCode()) {
+    SetError(MakeInternalError());
+    return;
+  }
+
+  absl::optional<SolanaAccountInfo> account_info;
+  if (!solana::ParseGetAccountInfo(api_request_result.body(), &account_info)) {
+    SetError(MakeInternalError());
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:509]: Should it be `SetError(ParseErrorResult(api_request_result.body()));` instead?

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+
+  if (!api_request_result.Is2XXResponseCode()) {
+    SetError(MakeInternalError());
+    return;
+  }
+
+  auto [parsing_ok, token_owner] =
+      GetTokenOwnerFromGetProgramAccountsResult(api_request_result.body());
+  if (!parsing_ok) {
+    SetError(MakeInternalError());
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:549]: Should it be `SetError(ParseErrorResult(api_request_result.body()));` instead?

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+    return;
+  }
+}
+
+void SnsResolverTask::FetchSolRecordRegistryState() {
+  DCHECK(!sol_record_name_registry_state_);
+
+  auto sol_record_address = GetDomainKey("SOL." + domain_, true);
+  if (!sol_record_address) {
+    // Put a domain owner as a fallback result and use it if there is no SOL
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:600]: Put the domain owner

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+  ScopedWorkOnSnsTask work_on_task(this);
+
+  if (!api_request_result.Is2XXResponseCode()) {
+    SetError(MakeInternalError());
+    return;
+  }
+
+  absl::optional<SolanaAccountInfo> account_info;
+  if (!solana::ParseGetAccountInfo(api_request_result.body(), &account_info)) {
+    SetError(MakeInternalError());
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:628]: `SetError(ParseErrorResult(api_request_result.body()));`?

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+  }
+
+  absl::optional<SolanaAccountInfo> account_info;
+  if (!solana::ParseGetAccountInfo(api_request_result.body(), &account_info)) {
+    SetError(ParseErrorResult(api_request_result.body()));
+    return;
+  }
+
+  if (!account_info) {
+    SetError(MakeInvalidParamsError());
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:583]: Should it be MakeInternalError?

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+    return;
+  }
+
+  domain_name_registry_state_ =
+      FromBase64<NameRegistryState>(account_info->data);
+  if (!domain_name_registry_state_) {
+    SetError(MakeInvalidParamsError());
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:590]: Should this be MakeInternalError?

**Review by yrliou — COMMENTED** [review]
I'll take another look at tests this afternoon.

File: components/brave_wallet/browser/json_rpc_response_parser.h [hunk 10]
```diff
@@ -80,6 +80,15 @@ absl::optional<std::string> ConvertMultiUint64InObjectArrayToString(
     const std::vector<std::string>& keys,
     const std::string& json);
 
+namespace solana {
+
+base::OnceCallback<absl::optional<std::string>(const std::string& raw_response)>
+ConverterForGetAccountInfo();
+base::OnceCallback<absl::optional<std::string>(const std::string& raw_response)>
+ConverterForGetProrgamAccounts();
```
Comment (yrliou) [inline components/brave_wallet/browser/json_rpc_response_parser.h:88]: How about putting in solana_response_parser.h instead?

File: components/brave_wallet/browser/sns_resolver_task.h [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+SnsNamehash GetHashedName(const std::string& name);
+
+absl::optional<SolanaAddress> GetNameAccountKey(
+    const SnsNamehash& hashed_name,
+    const absl::optional<SolanaAddress>& parent);
+
+absl::optional<SolanaAddress> GetMintAddress(
+    const SolanaAddress& domain_address);
+absl::optional<SolanaAddress> GetDomainKey(const std::string& domain,
+                                           bool record);
```
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.h:35]: Unit tests for these?

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+absl::optional<SolanaAddress> ParseAndVerifySolRecordData(
+    base::span<const uint8_t> sol_record_payload,
+    const SolanaAddress& sol_record_address,
+    const SolanaAddress& domain_owner) {
+  constexpr size_t kSolRecordDataSignature =
+      static_cast<size_t>(ED25519_SIGNATURE_LEN);
+  constexpr size_t kSolRecordDataSize =
+      kSolanaPubkeySize + kSolRecordDataSignature;
+
+  if (sol_record_payload.size() < kSolRecordDataSize) {
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:87]: This comes as 2k bytes vector with trailing zeros.

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+  auto sol_record_payload_signature =
+      sol_record_payload.subspan(kSolanaPubkeySize, kSolRecordDataSignature);
+
+  std::vector<uint8_t> message;
+  message.insert(message.end(), sol_record_payload_address->bytes().begin(),
+                 sol_record_payload_address->bytes().end());
+  message.insert(message.end(), sol_record_address.bytes().begin(),
+                 sol_record_address.bytes().end());
+
+  std::string hex_message = base::ToLowerASCII(base::HexEncode(message));
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:106]: https://github.com/Bonfida/solana-program-library/blob/171553544d76f5de294a0c041dfcb17834fe91c5/name-service/js/src/resolve.ts#L54

File: components/brave_wallet/browser/sns_resolver_task.h (hunk 11 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.h]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/json_rpc_response_parser.h (hunk 10 above)
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_response_parser.h]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 9 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 8 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 7 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 5 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/common/brave_wallet.mojom [hunk 14]
```diff
@@ -904,6 +904,9 @@ interface JsonRpcService {
   // ENS lookups
   EnsGetEthAddr(string domain, EnsOffchainLookupOptions? options) => (string address, bool require_offchain_consent, ProviderError error, string error_message);
 
+  // SNS lookups
+  SnsGetSolAddr(string domain) => (string address, ProviderError error, string error_message);
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/common/brave_wallet.mojom]: Didn't know about SolanaProviderError :)
https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 15]
```diff
@@ -0,0 +1,665 @@
...
+
+  // https://bonfida.github.io/solana-name-service-guide/domain-name/records.html#the-sol-record
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 16]
```diff
@@ -0,0 +1,665 @@
...
+        data_span.subspan(owner_offset, kSolanaPubkeySize));
+    if (!address)
+      return absl::nullopt;
+    result->owner = *address;
+    return result;
+  }
+};
+
+// Make getProgramAccounts RPC call to find token account for mint. Filters by
+// token account data size, nft amount eq to 1 and traget mint address.
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.h [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+  base::WeakPtrFactory<SnsResolverTask> weak_ptr_factory_{this};
+};
+
+template <typename ResultCallback>
+class SnsResolverTaskContainer {
+ public:
+  SnsResolverTaskContainer() = default;
+  SnsResolverTaskContainer(const SnsResolverTaskContainer&) = delete;
+  SnsResolverTaskContainer(SnsResolverTaskContainer&&) = delete;
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.h]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.h [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+  SnsResolverTaskResult(SnsResolverTaskResult&&);
+  SnsResolverTaskResult& operator=(const SnsResolverTaskResult&);
+  SnsResolverTaskResult& operator=(SnsResolverTaskResult&&);
+  ~SnsResolverTaskResult();
+
+  SolanaAddress resolved_address;
+};
+
+struct SnsResolverTaskError {
+  SnsResolverTaskError();
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.h]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.h [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+  SnsResolverTaskError(mojom::ProviderError error, std::string error_message);
+  SnsResolverTaskError(const SnsResolverTaskError&);
+  SnsResolverTaskError(SnsResolverTaskError&&);
+  SnsResolverTaskError& operator=(const SnsResolverTaskError&);
+  SnsResolverTaskError& operator=(SnsResolverTaskError&&);
+  ~SnsResolverTaskError();
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.h]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.h [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+  static absl::optional<NameRegistryState> FromBase64(const std::string& str);
+
+  SolanaAddress parent_name;
+  SolanaAddress owner;
+  SolanaAddress data_class;
+  std::vector<uint8_t> data;
+};
+
+struct SnsResolverTaskResult {
+  SnsResolverTaskResult();
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.h]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 21]
```diff
@@ -0,0 +1,665 @@
...
+  // Only interested in owner.
+  SolanaAddress owner;
+
+  static absl::optional<SplAccountData> FromBytes(
+      base::span<const uint8_t> data_span) {
+    // https://github.com/solana-labs/solana-program-library/blob/f97a3dc7cf0e6b8e346d473a8c9d02de7b213cfd/token/program/src/state.rs#L129
+    const size_t kSplAccountDataSize = 165;
+    absl::optional<SplAccountData> result;
+
+    if (data_span.size() < kSplAccountDataSize)
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: fixed https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 22]
```diff
@@ -0,0 +1,665 @@
...
+  // Only interested in supply.
+  uint64_t supply = 0;
+
+  static absl::optional<SplMintData> FromBytes(
+      base::span<const uint8_t> data_span) {
+    // https://github.com/solana-labs/solana-program-library/blob/f97a3dc7cf0e6b8e346d473a8c9d02de7b213cfd/token/program/src/state.rs#L37
+    const size_t kSplMintDataSize = 82;
+    absl::optional<SplMintData> result;
+
+    if (data_span.size() < kSplMintDataSize)
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: fixed v

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 23]
```diff
@@ -0,0 +1,665 @@
...
+  auto account_data = FromBase64<SplAccountData>(account_info->data);
+  if (!account_data)
+    return {false, absl::nullopt};
+
+  return {true, account_data->owner};
+}
+
+}  // namespace
+
+SnsNamehash GetHashedName(const std::string& name) {
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc [hunk 24]
```diff
@@ -0,0 +1,665 @@
...
+  // https://github.com/Bonfida/solana-program-library/blob/6e3be3eedad3a7f4a83c1b7cd5f17f89231e0bca/name-service/js/src/constants.ts#L13
+  constexpr char kHashPrefix[] = "SPL Name Service";
+  const std::string input = kHashPrefix + name;
+  return crypto::SHA256Hash(base::as_bytes(base::make_span(input)));
+}
+
+absl::optional<SolanaAddress> GetNameAccountKey(
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:237]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 1 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:271]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 2 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:293]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 3 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 4 above)
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15617/commits/1e34bd9363c2e1aa4593c00f3409b8c119ccab29

File: components/brave_wallet/browser/json_rpc_response_parser.h (hunk 10 above)
Comment (yrliou) [reply components/brave_wallet/browser/json_rpc_response_parser.h]: @supermassive Move ConverterForGetAccountInfo too.

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/sns_resolver_task.cc:88]: Ok got it, please put a comment here, thanks.

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/sns_resolver_task.cc:107]: Thanks.

**Comment by yrliou** [comment]
Please create and link an issue, and fill out the test plan, thanks.

File: components/brave_wallet/browser/json_rpc_response_parser.h (hunk 10 above)
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_response_parser.h]: https://github.com/brave/brave-core/pull/15617/commits/e2a76050ea88794a45f742328f6971a8c2694af8

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:111]: added as comment https://github.com/brave/brave-core/pull/15617/commits/e2a76050ea88794a45f742328f6971a8c2694af8

File: components/brave_wallet/browser/sns_resolver_task.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:90]: https://github.com/brave/brave-core/pull/15617/commits/e2a76050ea88794a45f742328f6971a8c2694af8

**Review by yrliou — APPROVED** [review]
Code LGTM, please create and link an issue, and fill the test plan before merging, thanks.

**Pull request merged**